### ISSUE TITLE
config set LABEL

### ISSUE THREAD

**Issue by buhtignew** [slimcoin-project/pacli#12]
I've tried to test the `config set` command but have not understood what should I use as label.
Tried `config set network` but got the following output:
```
  File "~/.local/bin/pacli", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/__main__.py", line 441, in main
    fire.Fire({
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/extended_classes.py", line 72, in set
    write_settings(label, value)
    ^^^^^^^^^^^^^^
NameError: name 'write_settings' is not defined
```
Are we speaking about the label that the traditional and the extended files have?
Are they already set up? (It doesn't seem to me to see the flag to set a new one, for instance.)
I've tried with `config set extended` but got the following error:
```
  File "~/.local/bin/pacli", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/__main__.py", line 441, in main
    fire.Fire({
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
                                ^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^
  File "~/.local/lib/python3.11/site-packages/pacli/extended_classes.py", line 70, in set
    raise ValueError({'error': 'Invalid setting key.'}) # ValueError added
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: {'error': 'Invalid setting key.'}
```

**Comment by d5000** [comment]
I see the error messages here aren't intuitive still.

Basically you can set a label for every category of the basic and the extended config file.

In the extended one, for example, to set a label for a deck:

`pacli config set my_deck 3369f3eb703656c0b46b613dcfb4df78fc05657a0b75f44551634de6321d9d9c deck -e`

To change the basic configuration file parameters, look which ones are in the pacli.conf configuration file, or use `config list`.
The command is designed in a way it's compatible with the original PeerAssets config set command, so it's not ideal from usability reasons, but I regard it as a power user command.

**Comment by buhtignew** [comment]
I'd like to keep this topic open for the moment, because I'm interested to test this command to understand how it works.

**Comment by buhtignew** [comment]
I've tested `config set testnet1 testnet -m` and got the following error:
```
Error: Invalid setting key. This key doesn't exist in the standard configuration file.
```
What I was doing wrong here?

**Comment by d5000** [comment]
> What I was doing wrong here?

Documentation was a bit misleading here (fixed it).
First, --modify does not work with the basic config file. The labels are static, i.e. you can only change the values. If you try modifying the labels instead, now an error message will be shown.
Second, there is no testnet setting in the basic configuration file. Here `config list` was misleading as it showed all the config settings (except those in the extended file), not only those defined in the basic file. There are some which are set after pacli starts, for example the testnet variable.

I have thus changed the default behaviour of `config list`. The default output is now that only the file contents are shown. If you want all basic settings, you have now to add `-a` or `--all_basic_settings` (the long flag name is to prevent people from thinking this will also show the extended file settings).

>  the -e flag was optional.

Confirmed, this is not really intended. This will be fixed finally when the treatment of positional arguments is fixed generally in pacli (See issue #15 ).

>  Would it make sense/be technically possible to make it optional as well in both cases?

It can still conflict in some cases with the command mode which edits the basic configuration file. I don't want it to be optional for this reason, so this current "optionality" has to be considered a bug. But as I wrote in #15  I have a plan to fix it, only that it has not the highest priority now.

**Comment by buhtignew** [comment]
> First, --modify does not work with the basic config file. The labels are static, i.e. you can only change the values. If you try modifying the labels instead, now an error message will be shown.

After I've read the following in the help:
```
    pacli config set LABEL VALUE [-e/--extended CATEGORY]

    Adds a setting in the basic configuration file (by default: pacli.conf)
    or a category CATEGORY in the extended configuration file.
```
I've tried `config set blabla True` and got the following message:
```
Error: Invalid setting key. This label doesn't exist in the standard configuration file. See permitted labels with: 'config list -l'.
```
However it seems like `config list -l` is not implemented yet since I'm receiving the following message while trying to run it:
```
ERROR: Could not consume arg: -l
```
There is not such flag mentioned in `config list -h` either.

So I've made another attempt trying to modify the `network` label value: `config set network tssl` and got the message that probably what you were referring to when you were saying above _If you try modifying the labels instead, now an error message will be shown._:
```
Error: Basic settings can only be modified with the --replace/-r flag. New labels can't be added.
```
I think we should modify this part in help making it clear that ` pacli config set LABEL VALUE [-e/--extended CATEGORY]` just can't be used for the basic configuration file, can it?

**Comment by d5000** [comment]
> And got that in case of basic file it doesn't actually replaces the value of the LABEL, but the value of VALUE (which we'll probably need to reformulate in this part).

Yes the formulation is still not ideal, but what I meant is that it replaces the value "associated with the LABEL". I'll look for a better wording.

> So I've tested config set rpcport 41688 -r and it was a perfect receipt for disaster because soon after that I was not able to use pacli because the connection error was thrown.

Exactly, that's also why I added the `-r` flag as mandatory (and wrote above that you should know what you're doing if using this command ;-) ). I may have to add more warnings to this option. An idea: to display a warning during let's say 20 seconds before the contents are modified, which could be overridden with a `--quiet` flag (for scripts).

> So I'd say the whole config set command is about working with extended file, except the possibility to modify the values in the basic file.

Yes. `config set` is originally a vanilla (original PeerAssets) command. Originally you could modify values of pacli.conf with `config set LABEL VALUE` without using the `--replace` flag. But I thought then that it would be better to require a flag to make it less likely that the user modifies something important like you did. And it's also consistent to how the command now works with the extended file.

Unfortunately I haven't found a way to run pacli without connection to a SLM node. This would be a nice addition, above all to deal with the config files (which does not require the connection), but it would require probably a modification of the original code when pacli is started. It's not really something prioritary now.

> At this point I'm wondering whether we can reformulate the whole help section but also suppress the -e flag

This may be an option. My approach to use the `-e` flag is based on the idea that the extended config can be modified with much more user-friendly variants (like `deck set`, `address set` etc.) and that `config set` is an original PeerAssets command which was to be used for the basic file only originally. I modified `config set` for the extended file only as a tool for advanced users, for example to fix errors in the config file, and thus I think entering a flag more isn't much of a problem.

So in general I don't think it's a real improvement to suppress `-e` and replace it by `-b` for the basic file - we would have to do this also in the other config commands.

**Comment by buhtignew** [comment]
> I may have to add more warnings to this option. An idea: to display a warning during let's say 20 seconds before the contents are modified, which could be overridden with a --quiet flag (for scripts).

I haven't used those commands in scripts for the moment without having an experience I can't say how useful it can be in that context.

However I think that by delaying the decision we are not communicating anything new to the user, on the opposite he may become annoyed with to much care about his fate we are putting into and begin doing things without considering the meaning of the warnings.

I feel that no basic file modification can be done without consequences for the correct usage of `pacli`. So or we should just prevent people from modifying anything in the basic file using that command or we should output the warning before the command is actually executed explaining what will happen after that value is modified in that way and how to solve the issue that may occur (this is a bit in contrast with the previous section :-) ) or we can make a mix of this two solutions.

> So in general I don't think it's a real improvement to suppress -e and replace it by -b for the basic file - we would have to do this also in the other config commands.

What I was trying to say is that the command can be used in 4 different ways and among all those ways only one is suitable for the basic file. This means to me that the command is more about working with extended file than with the basic file. So what is mentioned as optional in our help is actually mandatory. That's why I was suggesting us to consider suppressing the `-e` flag. And also if we consider the extended file as the main object the command is working on we'll probably be able to reduce the quantity of things to say in the help file.

The same thing with `config list`: if we consider the extended file content as default output it will simplify the understanding of this command for the advanced users. And the `config list` without flags would output the extended flag.

It can seem I'm insisting on something of not that much importance, but as user I feel it's yet a bit difficult to use our help for those two commands, especially that of `config set`.

**Comment by d5000** [comment]
>  we should output the warning before the command is actually executed explaining what will happen after that value is modified

That's what I meant. Of course while the warning is displayed nothing would be changed. There could be even a Y/N confirmation before. Also there will be a more drastic warning in the help itself.

>-e flag

I understand what you wrote, but my assesment on the topic is different.
As I wrote in the last paragraph the intended way to deal with the extended file for non-technical users is to use the dedicated commands for each category (`deck set` etc.).
`config set `is the only way to deal with the basic config file, and it's an original PeerAssets command.
I hope you understand but I'll not be losing more time with this.

**Comment by buhtignew** [comment]
> I hope you understand but I'll not be losing more time with this.

I was not trying to annoy you sorry if I did.
I'm only was seeing a huge simplifying opportunity in this command.
However since it will be used only by very advanced users we can also let it as it is.
I'm always open to describe you how I'm seeing the simplification here if necessary.

> That's what I meant. Of course while the warning is displayed nothing would be changed. There could be even a Y/N confirmation before. Also there will be a more drastic warning in the help itself.

I haven't thought about Y/N dialog. That's make sense.

**Comment by buhtignew** [comment]
> As I wrote in the last paragraph the intended way to deal with the extended file for non-technical users is to use the dedicated commands for each category (deck set etc.).
The reason why there are more options for the extended config file is because the extended file supports more options - you can add labels freely, delete settings, modify labels etc.. Not because this command is meant to be used more with the extended file.

After I've worked on [this](https://github.com/slimcoin-project/pacli/issues/15#issuecomment-1948527471) yesterday I've understood your point.
There is no need implementing any additional `-b` flag and/or "simplifying" `config set` command.

I think we can close this issue.

**Comment by d5000** [comment]
>  I'm always open to describe you how I'm seeing the simplification here if necessary.
Of course you can always propose simplifications. Only sometimes I think we discuss too long about a detail I consider not that relevent, and lose time with that I could spend into bugfixing. We should also in this stage not insist in a "perfect" CLI. If it's coherent and somewhat usable, it's ok for me. Coherency/internal "logic" is for me more important than finding the perfect flag keyword because a keyword can be changed later, but if the logic is wrong, correcting that is a major time sink.

I'll leave this issue open to remind me that I have to implement the Y/N dialog still, once this is done I'll close it.

**Comment by d5000** [comment]
Implemented Y/N dialog in commit 9e1f00c. So I close it.
